heapster-saw/src/Verifier/SAW/Heapster/Implication.hs

```diff
@@ -6226,37 +6284,23 @@ proveVarAtomicImpl x ps mb_p = case mbMatch mb_p of
   -- FIXME HERE: eventually we should handle lowned permissions on the right
   -- with arbitrary contained lifetimes, by equalizing the two sides
   [nuMP| Perm_LOwned [] _ _ |]
-    | [Perm_LOwned ls@(PExpr_Var l2:_) ps_in ps_out] <- ps ->
-      implEndLifetimeRecM l2 >>>
-      implRemoveContainedLifetimeM x ls ps_in ps_out l2 >>>
+    | [Perm_LOwned (PExpr_Var l2:_) _ _] <- ps ->
+      implPopM x (ValPerm_Conj ps) >>> implEndLifetimeRecM l2 >>>
       proveVarImplInt x (fmap ValPerm_Conj1 mb_p)
 
   [nuMP| Perm_LOwned [] mb_ps_inR mb_ps_outR |]
     | [Perm_LOwned [] ps_inL ps_outL] <- ps ->
 
-      -- First, simplify both sides using any current equality permissions. This
-      -- just builds the equality proofs and computes the new LHS and RHS, but
-      -- we don't actually perform the casts until later.
-      substEqsWithProof (ps_inL, ps_outL) >>>= \eqp_psL ->
-      get >>>= \s ->
-      give (cruCtxProxies (view implStateVars s))
-        (substEqsWithProof (mb_ps_inR, mb_ps_outR)) >>>= \eqp_mb_psR ->
-      let (ps_inL',ps_outL') = someEqProofRHS eqp_psL
-          (mb_ps_inR',mb_ps_outR') = someEqProofRHS eqp_mb_psR in
-
-      -- Pop ps from the stack, so we can push it to the top of the stack later
-      implPopM x (ValPerm_Conj ps) >>>
-
       -- Compute the necessary "permission subtractions" to figure out what
       -- additional permissions are needed to prove both ps_inR -o ps_inL and
-      -- ps_outL -o ps_outR. These required permissions are calls ps1 and ps2,
+      -- ps_outL -o ps_outR. These required permissions are called ps1 and ps2,
       -- respectively. Note that the RHS for both of these implications needs to
       -- be in a name-binding for the evars and the LHS needs to not be in a
       -- name-binding, so ps_inR cannot have any evars.
-      partialSubstForceM mb_ps_inR' "proveVarAtomicImpl" >>>= \ps_inR' ->
-      let mb_ps_inL' = fmap (const ps_inL') mb_ps_inR' in
-      solveForPermListImpl ps_inR' mb_ps_inL' >>>= \(Some neededs1) ->
-      solveForPermListImpl ps_outL' mb_ps_outR' >>>= \(Some neededs2) ->
+      partialSubstForceM mb_ps_inR "proveVarAtomicImpl" >>>= \ps_inR ->
+      let mb_ps_inL = fmap (const ps_inL) mb_ps_inR in
```

Comment:
```suggestion
      let mb_ps_inL = ps_inL <$ mb_ps_inR in
```

Reply:
Using some of these special-purpose operators sometimes just looks like line noise to me. I know `<$>` and `<*>`, but when it starts getting more complex I would rather just be verbose about what the operator is doing. It's just like using the lens library: if you don't know all those operators, it can be really hard to read code that uses them.